### PR TITLE
Simplifier: new interface for pointer-related simplifications

### PR DESCRIPTION
**Pull request by kroening**
Only look at last commit.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -2563,17 +2550,50 @@ bool simplify_exprt::simplify_node(exprt &expr)
     }
   }
   else if(expr.id()==ID_pointer_object)
-    no_change = simplify_pointer_object(expr) && no_change;
+  {
+    auto r = simplify_pointer_object(expr);
```
Comment (smowton) [inline src/util/simplify_expr.cpp:2554]: How about factoring this repeated code with a helper function:
```
static bool set_if_changed(exprt &expr, const simplify_resuttt<> &result)
{
  if(result.has_changed())
    expr = result.expr;
  return result.has_changed();
}
```

Then make all these into `changed = set_if_changed(expr, simplify_pointer_object(expr)) || changed;`

Comment (kroening) [reply src/util/simplify_expr.cpp:2554]: This will all collapse once all (or some majority) of the cases use the new interface!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b3b8b744b91a3aee4858babc63d259d2bcd79339).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117517651

**Review by peterschrammel — APPROVED** [review]

File: src/util/simplify_expr.cpp (hunk 1 above)
Comment (peterschrammel) [reply src/util/simplify_expr.cpp:2554]: > This will all collapse once all (or some majority) of the cases use the new interface!

Ok.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cec9dd8ec24ca481abbb888b1c1e8d281b7a3780).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117588700

**Pull request merged**